File: src/clients/javascript/sharpcoredb-client/examples/pooling-example.js
#!/usr/bin/env node

/**
 * Connection Pooling Example
 * Demonstrates efficient connection reuse for high-performance applications
 */

const { createPool } = require('../dist/index.js');

async function worker(workerId, pool, tasks) {
  const results = [];

  for (let i = 0; i < tasks; i++) {
    // Get connection from pool
    const connection = await pool.getConnection();

    try {
      // Simulate some work
      const result = await connection.execute('SELECT ? as worker, ? as task', {
        worker: workerId,
        task: i
      });
      results.push(result.rows[0]);

      // Simulate processing time
      await new Promise(resolve => setTimeout(resolve, 10));
    } finally {
      // Return connection to pool
      await connection.close();
    }
  }

  console.log(`👷 Worker ${workerId}: Completed ${tasks} tasks`);
  return results;
}

async function main() {
  console.log('🏊 @sharpcoredb/client Pooling Example');
  console.log('======================================');

  let pool;

  try {
    // Create a connection pool
    console.log('🏗️  Creating connection pool...');
    const poolConfig = {
      database: 'example',
      minConnections: 2,
      maxConnections: 10,
      maxIdleTime: 60000, // 1 minute
      acquireTimeout: 10000 // 10 seconds
    };

    pool = await createPool('localhost', 5001, poolConfig);
    console.log('✅ Pool created');
    console.log(`   Max connections: ${poolConfig.maxConnections}`);
    console.log(`   Min connections: ${poolConfig.minConnections}`);

    // Show initial stats
    const initialStats = pool.stats;
    console.log(`   Initial stats: ${JSON.stringify(initialStats)}`);

    // Run concurrent workers
    const numWorkers = 5;
    const tasksPerWorker = 10;
    console.log(`\n🚀 Starting ${numWorkers} workers with ${tasksPerWorker} tasks each...`);

    const startTime = Date.now();

    // Create worker tasks
    const workerPromises = [];
    for (let i = 0; i < numWorkers; i++) {
      workerPromises.push(worker(i, pool, tasksPerWorker));
    }

    // Run all workers concurrently
    const results = await Promise.all(workerPromises);

    const endTime = Date.now();
    const totalTime = endTime - startTime;
    const totalTasks = results.reduce((sum, workerResults) => sum + workerResults.length, 0);

    // Show final stats
    const finalStats = pool.stats;
    console.log(`\n📊 Final pool stats: ${JSON.stringify(finalStats)}`);

    console.log('\n🎯 Performance Results:');
    console.log(`   Total tasks completed: ${totalTasks}`);
    console.log(`   Total time: ${(totalTime / 1000).toFixed(2)}s`);
    console.log(`   Tasks/second: ${(totalTasks / (totalTime / 1000)).toFixed(1)}`);
    console.log(`   Avg latency: ~${((totalTime / totalTasks) + 10).toFixed(0)}ms per task`);

  } catch (error) {
    console.error('❌ Error:', error.message);
    process.exit(1);
  } finally {
    if (pool) {
      console.log('\n🔌 Closing pool...');
      await pool.close();
      console.log('✅ Pool closed');
    }
  }

  console.log('\n🎉 Pooling example completed successfully!');
}

// Run the example
main().catch(console.error);

File: src/clients/python/pysharpcoredb/examples/pooling_example.py
#!/usr/bin/env python3
"""Example: Connection pooling for high-performance applications."""

import asyncio
import time
import pysharpcoredb as scdb


async def worker(worker_id: int, pool: scdb.ConnectionPool, tasks: int):
    """Worker function that uses connections from the pool."""
    results = []

    for i in range(tasks):
        async with pool.get_connection() as conn:
            # Simulate some work
            result = await conn.execute("SELECT 1 as worker, ? as task", {"task": i})
            results.append(result.rows[0].values)
            await asyncio.sleep(0.01)  # Simulate processing time

    print(f"Worker {worker_id}: Completed {tasks} tasks")
    return results


async def main():
    """Demonstrate connection pooling with concurrent workers."""
    print("PySharpDB Connection Pooling Example")
    print("=" * 40)

    # Create a connection pool
    pool_config = {
        "host": "localhost",
        "port": 5001,
        "database": "example",
        "min_connections": 2,
        "max_connections": 10,
        "max_idle_time": 60.0,
        "acquire_timeout": 10.0
    }

    try:
        async with scdb.create_pool(**pool_config) as pool:
            print("✅ Created connection pool")
            print(f"   Max connections: {pool.max_connections}")
            print(f"   Min connections: {pool.min_connections}")

            # Show initial stats
            stats = pool.stats
            print(f"   Initial stats: {stats}")

            # Run concurrent workers
            num_workers = 5
            tasks_per_worker = 10

            print(f"\n🚀 Starting {num_workers} workers with {tasks_per_worker} tasks each...")

            start_time = time.time()

            # Create worker tasks
            tasks = [
                worker(i, pool, tasks_per_worker)
                for i in range(num_workers)
            ]

            # Run all workers concurrently
            results = await asyncio.gather(*tasks)

            end_time = time.time()

            # Show final stats
            final_stats = pool.stats
            print(f"\n📊 Final pool stats: {final_stats}")

            total_tasks = sum(len(worker_results) for worker_results in results)
            total_time = end_time - start_time

            print("\n🎯 Performance Results:")
            print(f"   Total tasks completed: {total_tasks}")
            print(f"   Total time: {total_time:.2f}s")
            print(f"   Tasks/second: {total_tasks / total_time:.1f}")
    except Exception as e:
        print(f"❌ Error: {e}")
        return 1

    print("\n🎉 Pooling example completed successfully!")
    return 0


if __name__ == "__main__":
    exit_code = asyncio.run(main())
    exit(exit_code)
